### ISSUE TITLE
Unreleased double singletons feature issues

### ISSUE THREAD

**Issue by bitwes** [bitwes/Gut#346]
The `double_singleton` and `partial_double_singleton` features have been disabled in 7.3.0 and the wiki documentation has been removed.  When trying to create a demo project using `Engine` and `OS` I found numerous issues that were not discovered in the initial testing using `Input`.  The feature was inspired by use with `Input`, but some of the other common uses were not explored fully. 

Issues found:
* In testing `Input` it appeared that all methods related to `Input` had a method id > 1000.  Everything less than 1000 was ignored.  This is not the case for `OS`.  The other classes were not examined further.  Changing the code to add every method > 200 seemed to fix `OS`, but without further investigation we can't be sure how doubles of other singletons will behave.
* It appears that a double of `Engine` did not get the `editor_hint` property.  I was able to stub the `is_editor_hint()` method, but directly using the property `editor_hint` resulted in an error.

This feature will have to be explored further before it can be released.  

The `double_singleton` and `partial_double_singleton` methods have been hardcoded to return `null` and tests have been commented out.  This was found at the last minute and I hope to re-introduce the feature in the next release.  If not, then I will remove the commented out stuff then.  Hooray tech-debt!

**Comment by lars-hoeck** [comment]
Hey, is this feature dead? I think it's a pretty important feature.

**Comment by bitwes** [comment]
This is not dead, but I haven't got around to reworking the problem.  It is still high on the priority list, but most of my efforts recently have been dealing with Godot 4.0.

**Comment by SavoVuksan** [comment]
Hi,

first of all thanks for this nice addon ^^.

I wanted to ask if there is some updates on this feature. I am currently working on an addon and in my code I have a check if the script is running in the editor or not and I noticed I couldn't double/mock the Engine Singleton.

I wanted to ask if it is possible to Double my own singleton? Because if so I could create my own singleton that has a runs_in_editor method that I could mock?

**Comment by jabsatz** [comment]
I assume this is the issue preventing doing something like `double(Time)` ? I'm new to this library so I'm not sure if I'm missing something, but if that's the case then being unable to mock stuff like timestamps is definitely unfortunate :(

**Comment by bitwes** [comment]
There is no way that I know of to double a singleton and make it automatically available to objects under test.  You must at least have a variable which is assigned to the singleton that can be replaced in tests.  Using `Time` as the example you can do the following:

```gdscript
class_name Foobar
var t = Time

func foo():
	return t.get_ticks_msec()
```

```gdscript
extends GutTest

class TimeMock:
	func get_ticks_msec():
		return 0

func before_all():
	# required to be able to double TimeMock
	register_inner_classes(get_script())

func test_foo_returns_msecs():
	var mock_time = double(TimeMock).new()
	stub(mock_time.get_ticks_msec).to_return(10)

	var f = Foobar.new()
	f.t = mock_time
	
	assert_eq(f.foo(), 10)
```

I think that I can leverage ClassDB to automate the creation of classes like `TimeMock`.  Unfortunately these mock classes cannot inherit from their Singleton sources (this is how all other doubles work).  It appears this test still works with `var t := Time`.  I think this is because `Time` extends `Object` and nothing else.  I'm not sure if this is true for all Godot singletons.

**Comment by jabsatz** [comment]
I just tried this solution and it perfectly fixed my problem, thank you very much. It might be worth adding this to the docs somewhere, since I think mocking time is one of the prime examples of methods you'd want to mock in a test.
I think I was pretty thorough looking for a solution everywhere but just never thought about making a global godot class into a variable for this, since it's not how I generally engage with them.
Regardless, thank you so much once again! I'm really liking this library :)

**Comment by bitwes** [comment]
> It might be worth adding this to the docs somewhere

I agree.  I thought I had put it somewhere but I couldn't find it either.

I started digging back into this and it looks promising.  It looks like I can get a list of the singleton methods, properties, and enums.  I should be able to make a special double and even partial double of singletons that punch through to the singleton for method calls.  Properties probably shouldn't punch through to Singleton...though it would be possible.

**Comment by bitwes** [comment]
Well, it's finally done.  The documentation I couldn't find was in a stashed away wiki page that was being excluded.  You can now create wrapper doubles of most of the singletons that Godot provides.

https://gut.readthedocs.io/en/latest/Doubling-Singletons.html
